src/commands/git-metadata/gitdb.ts

```diff
@@ -0,0 +1,297 @@
+    log.info('[unshallow] Setting remote.origin.partialclonefilter to "blob:none" to avoid fetching file content')
+    await git.addConfig('remote.origin.partialclonefilter', 'blob:none')
+    log.info('[unshallow] Running git fetch --shallow-since="1 month ago" --update-shallow --refetch')
+    await git.fetch(['--shallow-since="1 month ago"', '--update-shallow', '--refetch'])
+    log.info('[unshallow] Fetch completed.')
+  }
+}
+
+// getKnownCommits asks the backend which of the given commits are already known
+const getKnownCommits = async (log: Logger, request: RequestBuilder, repoURL: string, latestCommits: string[]) => {
+  interface SearchCommitResponse {
+    data: Commit[]
+  }
+
+  interface Commit {
+    type: string
+    id: string
+  }
+
+  const localCommitData = JSON.stringify({
+    meta: {
+      repository_url: repoURL,
+    },
+    data: latestCommits.map((commit) => ({
+      id: commit,
+      type: 'commit',
+    })),
+  })
+  const response = await runRequest(log, 'search_commits', () =>
+    request({
+      url: '/api/v2/git/repository/search_commits',
+      headers: {
+        'Content-Type': 'application/json',
+      },
+      data: localCommitData,
+      method: 'POST',
+      timeout: API_TIMEOUT,
+    })
+  )
+  const commits = response.data as SearchCommitResponse
+  if (!commits || commits.data === undefined) {
+    throw new Error(`Invalid API response: ${response}`)
+  }
+
+  return commits.data.map((c) => {
+    if (c.type !== 'commit' || c.id === undefined) {
+      throw new Error('Invalid commit type response')
+    }
+
+    return sanitizeCommit(c.id)
+  })
+}
+
+const sanitizeCommit = (sha: string) => {
+  const isValidSha = (s: string) => /[0-9a-f]{40}/.test(s)
+
+  const sanitizedCommit = sha.replace(/[^0-9a-f]+/g, '')
+  if (sanitizedCommit !== sha) {
+    throw new Error(`Invalid commit format: ${sha} (different from sanitized ${sanitizedCommit})`)
+  }
+  if (!isValidSha(sanitizedCommit)) {
+    throw new Error(`Invalid commit format: ${sanitizedCommit}`)
+  }
+
+  return sanitizedCommit
+}
+
+const getObjectsToUpload = async (git: simpleGit.SimpleGit, commitsToExclude: string[]) => {
+  const rawResponse = await git.raw(
+    ['rev-list', '--objects', '--no-object-names', '--filter=blob:none', '--since="1 month ago"', 'HEAD'].concat(
+      commitsToExclude.map((sha) => '^' + sha)
+    )
+  )
+  const commitsToInclude = rawResponse.split('\n').filter((c) => c !== '')
+
+  return commitsToInclude
+}
+
+const generatePackFilesForCommits = (log: Logger, commits: string[]) => {
+  if (commits.length <= 0) {
+    return []
+  }
+
+  const generatePackfiles = (packfilePath: string) => {
+    const packObjectResults = child_process
+      .execSync(`git pack-objects --compression=9 --max-pack-size=3m ${packfilePath}`, {
+        input: commits.join('\n'),
+      })
+      .toString()
+      .split('\n')
+      .filter((sha) => sha)
```

Comment:
To be honest not a clear what this filter does? Removes empty lines?

Reply:
Yes. I changed the filter to make it more clear.

Reply:
Ok, but what about other lines?
I tried to run `git pack-objects --compression=9 --max-pack-size=3m`
and I got something like
```
Enumerating objects: 1, done.
Counting objects: 100% (1/1), done.
ec27d75473692ce16b631211cf50995703274499
Writing objects: 100% (1/1), done.
```
maybe it depends on the git setting, but I don't see how other lines filtered there.

Reply:
Lines 1, 2 and 4 are written to stderr while line 3 is written to stdout. The command `execSync` only returns stdout.

Reply:
Ok! Thanks!